### PR TITLE
mds: account for header size during omap commit

### PR DESCRIPTION
**Pull request by vshankar**
fnode_t is set in the omap header during directory commit operation which isn't accounted when tracking operation size.

Could be a reason for the file system to become read-only.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
@lxbsz PTAL

**Comment by vshankar** [comment]
> @vshankar is there any reproducer for this? Or any logs?

I proposed this fix since we got a BZ report by RH QE and at that time no (debug) logs were available - so, this fix came entirely from code reading and that's the reason it is a draft PR. Now that I better understand the problem (which is entirely different) given that I was able to play around with the MDS which experienced this problem, this change is only needed when `mds_dir_max_commit_size` is equal to `osd_max_write_size`.

File: src/mds/CDir.cc [hunk 1]
```diff
@@ -2494,6 +2494,10 @@ void CDir::_omap_commit_ops(int r, int op_prio, int64_t metapool, version_t vers
       mdcache->mds->heartbeat_reset();
   }
 
+  // the last omap commit includes the omap header, so account for
+  // that size early on so that when we reach `commit_one(true)`,
+  // there is enough space for the header.
+  write_size += sizeof(fnode_t);
```
Comment (mchangir) [inline src/mds/CDir.cc:2500]: would this affect standby-replay path ?

Comment (vshankar) [reply src/mds/CDir.cc:2500]: I don't think so. OMAP commits are active MDS only.

**Comment by vshankar** [comment]
Converting this to ready-for-review. I'll have a tracker for this if the changes are acceptable.

**Review by mchangir — APPROVED** [review]
nice catch

**Comment by salieri11** [comment]
jenkins test windows

**Comment by salieri11** [comment]
Waiting for ceph management dashboard issue to be resolved and I will run TH tests on this branch. Will merge it after tests pass

**Comment by salieri11** [comment]
started shaman build

**Comment by salieri11** [comment]
Started another shaman build after rebasing, since last week we experienced issues with centos builds

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/68968](https://tracker.ceph.com/issues/68968).

**Comment by vshankar** [comment]
@salieri11 This build just fine in shaman. See: https://shaman.ceph.com/builds/ceph/wip-vshankar-testing-20241118.055430-debug/

What I tend to do is if the jenkins tests fail for unrelated reason, I still put the change to build and test and hope that the jenkins failures resolve by the time tests are done and verified. Saves a bit of time :)

**Comment by salieri11** [comment]
jenkins retest this please

**Comment by vshankar** [comment]
jenkins test submodules

**Comment by vshankar** [comment]
@salieri11 finally! :)

BTW - here is the fs suite run which includes this PR: https://pulpito.ceph.com/vshankar-2024-11-18_08:59:55-fs-wip-vshankar-testing-20241118.055430-debug-testing-default-smithi/

**Pull request merged**